stepflow-rs/crates/stepflow-server/src/api/batches.rs

```diff
@@ -0,0 +1,489 @@
+    /// List of outputs sorted by batch_input_index
+    pub outputs: Vec<BatchOutputInfo>,
+}
+
+/// Query parameters for listing batches
+#[derive(Debug, Clone, Deserialize, ToSchema, utoipa::IntoParams)]
+#[serde(rename_all = "camelCase")]
+pub struct ListBatchesQuery {
+    /// Filter by batch status
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub status: Option<BatchStatus>,
+    /// Filter by flow name
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub flow_name: Option<String>,
+    /// Maximum number of results to return
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub limit: Option<usize>,
+    /// Number of results to skip
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub offset: Option<usize>,
+}
+
+/// Create and execute a batch of flows
+#[utoipa::path(
+    post,
+    path = "/batches",
+    request_body = CreateBatchRequest,
+    responses(
+        (status = 200, description = "Batch created successfully and execution started", body = CreateBatchResponse),
+        (status = 400, description = "Invalid request"),
+        (status = 404, description = "Flow not found"),
+        (status = 500, description = "Internal server error")
+    ),
+    tag = crate::api::BATCH_TAG,
+)]
+pub async fn create_batch(
+    State(executor): State<Arc<StepflowExecutor>>,
+    Json(req): Json<CreateBatchRequest>,
+) -> Result<Json<CreateBatchResponse>, ErrorResponse> {
+    let batch_id = Uuid::new_v4();
+    let state_store = executor.state_store();
+
+    // Get the flow from the state store
+    let flow = state_store
+        .get_flow(&req.flow_id)
+        .await?
+        .ok_or_else(|| error_stack::report!(ServerError::WorkflowNotFound(req.flow_id.clone())))?;
+
+    let total_runs = req.inputs.len();
+    let max_concurrency = req.max_concurrency.unwrap_or(total_runs);
+
+    // Create batch record
+    state_store
+        .create_batch(batch_id, req.flow_id.clone(), flow.name(), total_runs)
+        .await?;
+
+    // Create run records and collect (run_id, input, index) tuples
+    let mut run_inputs = Vec::with_capacity(total_runs);
+    for (idx, input) in req.inputs.into_iter().enumerate() {
+        let run_id = Uuid::new_v4();
+
+        // Create run record
+        state_store
+            .create_run(
+                run_id,
+                req.flow_id.clone(),
+                flow.name(),
+                None, // No flow label for batch execution
+                false, // Batch runs are not in debug mode
+                input.clone(),
+            )
+            .await?;
+
+        // Link run to batch
+        state_store
+            .add_run_to_batch(batch_id, run_id, idx)
+            .await?;
+
+        run_inputs.push((run_id, input, idx));
+    }
+
+    // Spawn background task for batch execution
+    let executor_clone = executor.clone();
+    let flow_clone = flow.clone();
+    let flow_id = req.flow_id.clone();
+
+    tokio::spawn(async move {
```

Comment:
Would be good to split out the batch execution logic into a separate executor so we can more intelligently manage things.

Reply:
Opened #347.